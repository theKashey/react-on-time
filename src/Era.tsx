import * as React from 'react';
import * as PropTypes from 'prop-types';

export interface IComponentProps {
  onStart?: () => any
  onEnd?: () => any;
}

export class Era extends React.Component<IComponentProps> {
  static propTypes = {
    onStart: PropTypes.func,
    onEnd: PropTypes.func
  };

  componentDidMount() {
    if (this.props.onStart) {
      this.props.onStart();
    }
  }

  componentWillUnmount() {
    if (this.props.onEnd) {
      this.props.onEnd();
    }
  }

  render() {
    return (
      this.props.children
        ? <React.Fragment>{this.props.children}</React.Fragment>
        : null
    )
  }
}