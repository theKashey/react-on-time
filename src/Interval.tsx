import * as React from 'react';
import {Component} from 'react';

export interface ComponentProps {
  delay: number;
  pause?: boolean;
  onTick: (tick: number) => any;
  children?: (tick: number) => React.ReactNode;
}

export interface ComponentState {
  tick: number
}

export class Interval extends Component<ComponentProps, ComponentState> {
  state = {
    tick: 0
  };

  timeout: number;

  componentDidMount() {
    this.start()
  }

  componentDidUpdate(prevProps:ComponentProps) {
    if (prevProps.delay !== this.props.delay || prevProps.pause !== this.props.pause) {
      this.stop();
      this.start();
    }
  }

  componentWillUnmount() {
    this.stop()
  }

  start() {
    if (!this.props.pause) {
      this.timeout = window.setInterval(() => {
        this.setState(({tick}) => ({tick: tick + 1}), () =>
          this.props.onTick && this.props.onTick(this.state.tick)
        )
      }, this.props.delay || 1000);
    }
  }

  stop() {
    this.timeout && window.clearInterval(this.timeout);
  }

  render() {
    return this.props.children
      ? this.props.children(this.state.tick)
      : null;
  }
}