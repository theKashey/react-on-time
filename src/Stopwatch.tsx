import * as React from 'react';
import * as PropTypes from 'prop-types';

export interface IComponentProps {
  timeout: number;
  onTick: (progress: number) => any;
  children?: (progress: number) => React.ReactNode;
}

export interface IComponentState {
  tick: number
}

export class Stopwatch extends React.Component<IComponentProps, IComponentState> {
  static propTypes = {
    timeout: PropTypes.number.isRequired,
    onTick: PropTypes.func,
    children: PropTypes.func
  };

  state = {
    tick: 0
  };

  private raf: number;
  private timeMount: number;

  componentDidMount() {
    this.timeMount = Date.now();
    this.tick()
  }

  componentWillUnmount() {
    this.stop()
  }

  tick() {
    if ((Date.now() - this.timeMount) < this.props.timeout) {
      this.raf = requestAnimationFrame(this.onTick);
    }
  }

  stop() {
    cancelAnimationFrame(this.raf);
  }

  onTick = () => {
    const {onTick} = this.props;
    if (onTick) {
      onTick(this.getProgress());
    }
    this.setState({tick: this.state.tick + 1}, this.tick);
  };

  getProgress() {
    return Math.min(1, (Date.now() - this.timeMount) / this.props.timeout);
  }

  render() {
    return this.props.children
      ? this.props.children(this.getProgress())
      : null;
  }
}