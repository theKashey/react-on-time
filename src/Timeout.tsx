import * as React from 'react';
import {Component} from 'react';

export interface ComponentProps {
  timeout: number;
  then?: () => any;
  children?: (timedout: boolean) => React.ReactNode;
}

export interface ComponentState {
  timedout: boolean
}

export class Timeout extends Component<ComponentProps, ComponentState> {
  state = {
    timedout: false
  };

  timeout: number;

  componentDidMount() {
    this.timeout = window.setTimeout(() => {
      this.setState({timedout: true});
      this.props.then && this.props.then();
    }, this.props.timeout || 1000);
  }

  componentWillUnmount() {
    window.clearTimeout(this.timeout);
  }

  render() {
    return this.props.children
      ? this.props.children(this.state.timedout)
      : null;
  }
}