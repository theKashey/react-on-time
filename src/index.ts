import {Timeout} from "./Timeout";
import {Interval} from "./Interval";
import {Stopwatch} from "./Stopwatch";
import {Era} from './Era'

export {
  Timeout,
  Interval,
  Stopwatch,
  Era
}