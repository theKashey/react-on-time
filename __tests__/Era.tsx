import * as React from 'react';
import {mount} from 'enzyme';
import {Era} from "../src";

describe('Era', () => {
  it('Call callbacks on mount', () => {
    const spy = jest.fn();
    const wrapper = mount(<Era onStart={spy}/>);
    expect(spy).toHaveBeenCalled();
  });

  it('Call callbacks on unmount', () => {
    const spy = jest.fn();
    const wrapper = mount(<Era onEnd={spy}/>);
    wrapper.unmount();
    expect(spy).toHaveBeenCalled();
  });

  it('Renders children', () => {
    const wrapper = mount(<Era><span>1</span><span>2</span></Era>);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
