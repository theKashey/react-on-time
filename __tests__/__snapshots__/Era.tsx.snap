// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Era Renders children 1`] = `"<span>1</span>"`;
